### PR TITLE
add route level - redis caching 

### PR DESCRIPTION
**Pull request by NakulManchanda**
## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/3434

## Description of Changes
- Adds FIXME widget to TODO page.
- 

## Test Plan
- Unit tested the `FIXME()` call.
- CA (click around) tested on local and frack:
  - TODO page
  - 

## Deployment Plan

1. 

## Other Considerations

-

### CODE REVIEW

**Comment by NakulManchanda** [comment]
Steps to hit middleware:

Login>select any chain
/api/bulkOffchain?chain=dydx&community=&jwt=

Query Generated Without topics:
```
SELECT "Chain"."id", "Chain"."chain_node_id", "Chain"."name", "Chain"."discord_config_id", "Chain"."description", "Chain"."token_name",
 "Chain"."website", "Chain"."discord", "Chain"."element", "Chain"."telegram",
  "Chain"."github", "Chain"."default_symbol", "Chain"."network", "Chain"."base",
   "Chain"."ss58_prefix", "Chain"."icon_url", "Chain"."active", "Chain"."stages_enabled",
    "Chain"."custom_stages", "Chain"."custom_domain", "Chain"."block_explorer_ids", 
    "Chain"."collapsed_on_homepage", "Chain"."type", "Chain"."substrate_spec", 
    "Chain"."has_chain_events_listener", "Chain"."default_summary_view",
     "Chain"."default_page", "Chain"."has_homepage", "Chain"."hide_projects",
      "Chain"."terms", "Chain"."bech32_prefix", "Chain"."admin_only_polling", "Chain"."default_allow_permissions", 
      "Chain"."default_deny_permissions", "ChainNode"."id" AS "ChainNode.id", "ChainNode"."url" AS "ChainNode.url", "ChainNode"."eth_chain_id" AS "ChainNode.eth_chain_id", "ChainNode"."alt_wallet_url" AS "ChainNode.alt_wallet_url",
       "ChainNode"."balance_type" AS "ChainNode.balance_type", "ChainNode"."name" AS "ChainNode.name", "ChainNode"."description" AS "ChainNode.description", "ChainNode"."ss58" AS "ChainNode.ss58", "ChainNode"."bech32" AS "ChainNode.bech32", "ChainNode"."created_at" AS "ChainNode.created_at", "ChainNode"."updated_at" AS "ChainNode.updated_at" 
       FROM "Chains" AS "Chain" 
       INNER JOIN "ChainNodes" AS "ChainNode" ON "Chain"."chain_node_id" = "ChainNode"."id" 
       WHERE "Chain"."id" = 'dydx';
       ```

**Comment by NakulManchanda** [comment]
With topics
```
SELECT "Chain"."id", "Chain"."chain_node_id", "Chain"."name", "Chain"."discord_config_id", 
       "Chain"."description", "Chain"."token_name", "Chain"."website", "Chain"."discord", 
       "Chain"."element", "Chain"."telegram", "Chain"."github", "Chain"."default_symbol", 
       "Chain"."network", "Chain"."base", "Chain"."ss58_prefix", "Chain"."icon_url", "Chain"."active", 
       "Chain"."stages_enabled", "Chain"."custom_stages", "Chain"."custom_domain", "Chain"."block_explorer_ids", 
       "Chain"."collapsed_on_homepage", "Chain"."type", "Chain"."substrate_spec", "Chain"."has_chain_events_listener", 
       "Chain"."default_summary_view", "Chain"."default_page", "Chain"."has_homepage", "Chain"."hide_projects", 
       "Chain"."terms", "Chain"."bech32_prefix", "Chain"."admin_only_polling", "Chain"."default_allow_permissions", 
       "Chain"."default_deny_permissions", "topics"."id" AS "topics.id", "topics"."name" AS "topics.name", 
       "topics"."description" AS "topics.description", "topics"."telegram" AS "topics.telegram", 
       "topics"."chain_id" AS "topics.chain_id", "topics"."token_threshold" AS "topics.token_threshold", 
       "topics"."featured_in_sidebar" AS "topics.featured_in_sidebar", "topics"."featured_in_new_post" AS "topics.featured_in_new_post", 
       "topics"."order" AS "topics.order", "topics"."default_offchain_template" AS "topics.default_offchain_template", 
       "topics"."rule_id" AS "topics.rule_id", "ChainNode"."id" AS "ChainNode.id", "ChainNode"."url" AS "ChainNode.url", 
       "ChainNode"."eth_chain_id" AS "ChainNode.eth_chain_id", "ChainNode"."alt_wallet_url" AS "ChainNode.alt_wallet_url", 
       "ChainNode"."balance_type" AS "ChainNode.balance_type", "ChainNode"."name" AS "ChainNode.name", 
       "ChainNode"."description" AS "ChainNode.description", "ChainNode"."ss58" AS "ChainNode.ss58", 
       "ChainNode"."bech32" AS "ChainNode.bech32", "ChainNode"."created_at" AS "ChainNode.created_at", 
       "ChainNode"."updated_at" AS "ChainNode.updated_at" 
FROM "Chains" AS "Chain" 
LEFT OUTER JOIN "Topics" AS "topics" ON "Chain"."id" = "topics"."chain_id" AND ("topics"."deleted_at" IS NULL) 
INNER JOIN "ChainNodes" AS "ChainNode" ON "Chain"."chain_node_id" = "ChainNode"."id" 
WHERE "Chain"."active" = true;
```

**Comment by NakulManchanda** [comment]
@jake Redis caching
this is the direction it is going
https://github.com/hicommonwealth/commonwealth/pull/3492
router.get('/status', cacheDecorator.cache(60), status.bind(this, models));

- this is just router based cache
- will do query based caching in same PR to target queries instead of routes

sometimes its better to cache query because of parameters and sometimes its better to cache route especially which don't have params!

still investigating & iterating

File: packages/commonwealth/server.ts [hunk 1]
```diff
@@ -205,6 +205,10 @@ async function main() {
     app.use(passport.initialize());
     app.use(passport.session());
     app.use(prerenderNode.set('prerenderServiceUrl', 'http://localhost:3000'));
+    app.use(function (err, req, res, next) {
```
Comment (jnaviask) [inline packages/commonwealth/server.ts:208]: Why are we adding this extra middleware? We already have an error handling middleware in `setupErrorHandlers`.

File: packages/commonwealth/server/routing/router.ts [hunk 2]
```diff
@@ -214,7 +216,12 @@ function setupRouter(
     updateAddress.bind(this, models)
   );
   router.get('/domain', domain.bind(this, models));
-  router.get('/status', status.bind(this, models));
+  router.get('/status', cacheDecorator.cache(60, defaultUserKeyGenerator), status.bind(this, models));
+  router.get(
```
Comment (jnaviask) [inline packages/commonwealth/server/routing/router.ts:220]: Is this testing code? Shouldn't we ensure that testing code is not live in production?

File: packages/commonwealth/server/routing/router.ts [hunk 3]
```diff
@@ -214,7 +216,12 @@ function setupRouter(
     updateAddress.bind(this, models)
   );
   router.get('/domain', domain.bind(this, models));
-  router.get('/status', status.bind(this, models));
+  router.get('/status', cacheDecorator.cache(60, defaultUserKeyGenerator), status.bind(this, models));
```
Comment (jnaviask) [inline packages/commonwealth/server/routing/router.ts:219]: Are we really sure we want to cache the results of status? What cases did we verify correctness against?

File: packages/common-common/src/redisCache.ts [hunk 4]
```diff
@@ -120,18 +125,25 @@ export class RedisCache {
...
         'Redis client is not initialized. Run RedisCache.init() first!'
       );
       return false;
     }
-    const finalKey = namespace + '_' + key;
+    const finalKey = RedisCache.getNamespaceKey(namespace, key);
 
     try {
-      await this.client.set(finalKey, value);
+      if(duration>0) {
```
Comment (jnaviask) [inline packages/common-common/src/redisCache.ts:140]: Add spacing / lint.

File: packages/common-common/src/cacheDecorator.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        // Try to fetch the response from Redis cache
+        const cachedResponse = await this.redisCache.getKey(namespace, cacheKey);
+        if (cachedResponse) {
+          // Response found in cache, send it
+          console.log('Response found in cache, sending it');
+          res.set('X-Cache', 'HIT');
+          res.status(200).send(JSON.parse(cachedResponse));
+          return;
+        }
+        // Response not found in cache, generate it and cache it
```
Comment (jnaviask) [inline packages/common-common/src/cacheDecorator.ts:31]: Doubled comment?

File: packages/common-common/src/cacheDecorator.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        // Response not found in cache, generate it and cache it
+        const originalSend = res.send;
+        res.send = (body) => {
+          try {
+            console.log('Response not found in cache, sending it');
+            const response = originalSend.call(res,body);
+            try {
+              // const jsonBody = JSON.parse(body);
```
Comment (jnaviask) [inline packages/common-common/src/cacheDecorator.ts:39]: Dead code?

File: packages/common-common/src/cacheDecorator.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+              // if(jsonBody && jsonBody.status=='Success') {
+              if(res.statusCode == 200) {
```
Comment (jnaviask) [inline packages/common-common/src/cacheDecorator.ts:41]: Spacing / linting.

File: packages/common-common/src/cacheDecorator.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                this.redisCache.setKey(namespace, cacheKey, body, duration);
+              }
+            } catch (error) {
```
Comment (jnaviask) [inline packages/common-common/src/cacheDecorator.ts:44]: Is there any way to clean up this control flow? The triple nested catch block is a bit hard to follow in terms of how the middleware will function and what the possible failure cases are.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Added comments. Also, it seems like this PR doesn't actually do what it declares on the title, specifically "add new validateChainWithTopics middleware" -- it does something else entirely, and scope of work is not clear. Would love to get this PR properly documented so we can have some confidence on what changes are actually included before we merge.

File: packages/commonwealth/server.ts (hunk 1 above)
Comment (NakulManchanda) [reply packages/commonwealth/server.ts:208]: removing this one! left as debugging statement - we already set this up in setupErrorHandlers!

Comment (NakulManchanda) [reply packages/commonwealth/server.ts]: removed

File: packages/commonwealth/server/routing/router.ts (hunk 3 above)
Comment (NakulManchanda) [reply packages/commonwealth/server/routing/router.ts:219]: > do we want to?
- yes, its the second most frequent route
Basically, we want to cache these findAlls

https://github.com/hicommonwealth/commonwealth/blob/e89ae04f2b2f4db8fdf8d3b431ab2e309fa57062/packages/commonwealth/server/routes/status.ts#L72
```
      models.Chain.findAll({
      models.ChainNode.findAll(),
      models.NotificationCategory.findAll(),
      models.ChainCategory.findAll(),
      models.ChainCategoryType.findAll(),
```
- is it safe? 
- Looked inside the function - no useractivity updates in status function 
- only thing seems might frequently changed is `starrred communitites`
- its done at user level right now - `route_response_96271-/api/status`

File: packages/common-common/src/cacheDecorator.ts (hunk 8 above)
Comment (NakulManchanda) [reply packages/common-common/src/cacheDecorator.ts:44]: sorry, can try to take out this function as seperate one for readibility
- all try catch are different though
- first if some error in fetching from cache - just proceed to next middleware
- second returning erred response to user from intercepted res.send call
- third if some error writing to cache - just log and ignore

File: packages/common-common/src/cacheDecorator.ts (hunk 7 above)
Comment (NakulManchanda) [reply packages/common-common/src/cacheDecorator.ts]: made it `if(res.statusCode == 200) {`

not sure why liniting error not showing up in automated PR checks though!

File: packages/common-common/src/cacheDecorator.ts (hunk 8 above)
Comment (NakulManchanda) [reply packages/common-common/src/cacheDecorator.ts]: trying this out:
https://github.com/hicommonwealth/commonwealth/pull/3492/commits/3f373b0e88de7b3590eac04bf047437f15acc47a

File: packages/commonwealth/server/routing/router.ts (hunk 3 above)
Comment (jnaviask) [reply packages/commonwealth/server/routing/router.ts:219]: Is redis caching of the entire route's response the right solution when we only want to cache a part of the response?

File: packages/common-common/src/cacheDecorator.ts (hunk 7 above)
Comment (jnaviask) [reply packages/common-common/src/cacheDecorator.ts]: The proper formatting adds a space after the `if`, per convention across the codebase.

File: packages/common-common/src/redisCache.ts (hunk 4 above)
Comment (NakulManchanda) [reply packages/common-common/src/redisCache.ts]: hope looks better now!

File: packages/common-common/src/cacheDecorator.ts (hunk 5 above)
Comment (NakulManchanda) [reply packages/common-common/src/cacheDecorator.ts]: removed

File: packages/common-common/src/cacheDecorator.ts (hunk 7 above)
Comment (NakulManchanda) [reply packages/common-common/src/cacheDecorator.ts]: added!

File: packages/commonwealth/server/routing/router.ts (hunk 3 above)
Comment (NakulManchanda) [reply packages/commonwealth/server/routing/router.ts]: right place to do partial caching is inside the function, added function which can help with more customized caching.
- checkCacheAndSendResponseIfFound
- checkCache
- cacheResponse

In next PR -> i will caching database query which will take of that, `findAll` responses will populate model from cache instead of database.

**Comment by NakulManchanda** [comment]
> PR properly documented so we can have some confidence on what changes are actually included before we merge.
- Adding wiki for documentation
- PR just added some base tooling to help with caching, instead of making changes to app
- its very easy **to not use this PR**, as we made minimal changes to existing code
`router.get('/status', cacheDecorator.cacheMiddleware(60, defaultUserKeyGenerator), status.bind(this, models));` 
Just remove the middleware if found faulty
- we have to try it out on Frick for sure!

**Comment by NakulManchanda** [comment]
@jnaviask much more pleasant to read now. Still doing some local testing but feel free to give second look. Thanks!

File: packages/commonwealth/server/routing/router.ts (hunk 3 above)
Comment (jnaviask) [reply packages/commonwealth/server/routing/router.ts]: So what you're saying is we've not implemented the right solution for caching in the /status route? In that case, why are we moving forward with the wrong solution? There are certainly routes where redis caching of the entire response is helpful, but my understanding from this conversation is that /status is not one of them.

**Comment by jnaviask** [comment]
@NakulManchanda where is it documented how to spin up Redis locally for testing? I would like to validate locally before pushing to frick or frack.

**Comment by NakulManchanda** [comment]
i tried to add docker-compose earlier,
anyways i will add helper script like rabbitmq to spin up redis

File: packages/commonwealth/server/routing/router.ts (hunk 3 above)
Comment (NakulManchanda) [reply packages/commonwealth/server/routing/router.ts]: - for `/status` it will be cached per user, not as global response for every user because of user specific info
`route_response_user:1379_/api/status` - with TTL of 60secs
- when application bootstrapped it called without user as well, when user is logged, all those user will receive response from `route_response_/api/status`

**Comment by jnaviask** [comment]
Can't get redis cache to initialize locally. What env vars do I need?

**Comment by NakulManchanda** [comment]
> Can't get redis cache to initialize locally. What env vars do I need?

try adding this to .env
```
REDIS_URL=redis://localhost:6379
```

File: packages/common-common/src/cacheKeyUtils.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { Request } from 'express';
+
+export const defaultKeyGenerator = (req: Request) => {
+  return req.originalUrl;
+}
+
+export const defaultUserKeyGenerator = (req: Request) => {
```
Comment (NakulManchanda) [inline packages/common-common/src/cacheKeyUtils.ts:7]: this is how user level key is generated @jnaviask

File: packages/commonwealth/test/integration/api/cacheDecorator.spec.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      .get('/api/statusBroken')
+      .set('Accept', 'application/json');
+
+    // check redis - no key should be set
+    const valFromRedis = await redisCache.getKey(
+      route_namespace,
+      '/api/statusBroken'
+    );
+    expect(valFromRedis).to.be.null;
+    expect(res).to.have.status(500);
```
Comment (mhagel) [inline packages/commonwealth/test/integration/api/cacheDecorator.spec.ts:76]: should this be `res2`?

Comment (NakulManchanda) [reply packages/commonwealth/test/integration/api/cacheDecorator.spec.ts:76]: Yes

**Review by jnaviask — APPROVED** [review]
G2G -- no functional changes (yet)!
